networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/client_nsx.py

```diff
@@ -0,0 +1,206 @@
+                    LOG.error("Request={} Response={}".format(requestInfo,
+                                                              last_err))
+
+                msg = pattern.format(attempt, until, pause, method)
+
+                LOG.debug(msg)
+                eventlet.sleep(pause)
+
+            raise Exception(msg, last_err)
+
+        return decorator
+
+
+
+class Client:
+
+    def __init__(self):
+        rate=cfg.CONF.NSXV3.nsxv3_requests_per_second
+        timeout=cfg.CONF.NSXV3.nsxv3_requests_per_second_timeout
+        limit=1
+
+        self._api_scheduler = Scheduler(rate, limit, timeout)
+
+        self._timeout = cfg.CONF.NSXV3.nsxv3_request_timeout
+
+        self._base_path = 'https://{}:{}'.format(
+            cfg.CONF.NSXV3.nsxv3_login_hostname,
+            cfg.CONF.NSXV3.nsxv3_login_port
+        )
+
+        self._login_timestamp = 0
+        self._login_path = "/api/session/create"
+        self._login_data = {
+            "j_username": cfg.CONF.NSXV3.nsxv3_login_user,
+            "j_password": cfg.CONF.NSXV3.nsxv3_login_password
+        }
+
+        self._session = requests.session()
+
+        if cfg.CONF.NSXV3.nsxv3_suppress_ssl_wornings:
+            self._session.verify = False
+            requests.packages.urllib3.disable_warnings()
+
+        self._version = None
+
+    @property
+    def version(self, refresh=False):
+        if not self._version or refresh:
+            resp = self.get(path="/api/v1/node/version")
+            if resp.ok:
+                self._version = resp.json()['product_version']
+        return versionutils.convert_version_to_tuple(self._version)
+
+    def _login(self):
+        LOG.info("Session token - acquiring")
+        now = int(datetime.datetime.now().strftime("%s"))
+        with LockManager.get_lock(self._base_path):
+            if now > self._login_timestamp:
+                resp = requests.post(**self._params(path=self._login_path,
+                                                    data=self._login_data,
+                                                    verify=self._session.verify))
+                if resp.status_code != requests.codes.ok:
+                    resp.raise_for_status()
+
+                self._session.headers["Cookie"] = \
+                    resp.headers.get("Set-Cookie")
+                self._session.headers["X-XSRF-TOKEN"] = \
+                    resp.headers.get("X-XSRF-TOKEN")
+                self._session.headers["Accept"] = "application/json"
+                self._session.headers["Content-Type"] = "application/json"
+
+                self._login_timestamp = \
+                    int(datetime.datetime.now().strftime("%s"))
+
+        try:
+            # Refresh version after login
+            self.version(refresh=True)
+        except Exception:
+            pass
+        LOG.info("Session token - acquired, connected to NSX-T {}".format(self._version))
+
+    def _params(self, **kwargs):
+        kwargs["timeout"] = self._timeout
```

Comment:
What's the reason for setting `timeout` on every request instead of on the session?

Reply:
I read about issues with timeout on session objects, did not get deep into actually testing them, thus took the safe way.

Reply:
@joker-at-work I did not found a way to set this on a session, could you point me out?

Reply:
[Looks like](https://findwork.dev/blog/advanced-usage-python-requests-timeouts-retries-hooks/#setting-default-timeouts) you need to extend the `HTTPAdapter` to change the default behavior. Might be easier not to do this. Sorry, didn't remember that correctly.